### PR TITLE
rtnetlink: perform marshaling round-trip in RouteMessage tests

### PR DESCRIPTION
**Pull request by mdlayher**
Signed-off-by: Matt Layher <mlayher@fastly.com>

The marshaling tests are now comprehensive and also unmarshal to verify the structures are identical. The second set of tests for unmarshaling only now only deals with error cases.

/cc @djosephsen

### CODE REVIEW

**Comment by jsimonetti** [comment]
@mdlayher Without trying to sound like a douche, do you think this will tuely test rtnetlink or just the netlink.AttributeEncoder/Decoder implementation?
I have a feeling that if we stop providing binary input for the decoder, we are no longer testing actual rtnetlink messages, but only the encoder/decoder implementation.

What's your thought on this?

**Comment by mdlayher** [comment]
I don't think I am just testing the encoder/decoder. Notice that I am comparing that the marshaled output bytes exactly match the bytes specified in the test fixture.

It's important to be able to both encode and decode (and round-trip between the two with identical input/output) since they are two different code paths with big switch cases that need to be correct in order to send something sensible to the kernel.

If you want the unmarshaling step first to ensure we can consume raw bytes, I could do:

- unmarshal from test bytes
- compare against test struct
- marshal to bytes
- compare against test bytes
- unmarshal again
- compare against test struct

This way, all our bases are covered.

So basically I'd just add the unmarshaling step first and then verify that we can successfully round trip. Both the struct and bytes from the test fixture will be used to ensure consistent output.

**Comment by jsimonetti** [comment]
> So basically I'd just add the unmarshaling step first and then verify that we can successfully round trip. Both the struct and bytes from the test fixture will be used to ensure consistent output.

That sounds perfect. I think this will also give us the most flexibility to test for regressions should the need arise.

**Comment by jsimonetti** [comment]
Awesome, tnx!
